pkg/gadgets/utils/test/runner.go

```diff
@@ -0,0 +1,183 @@
+	UID int
+}
+
+type Info struct {
+	Pid       int
+	Comm      string
+	UID       int
+	MountNsID uint64
+}
+
+// Runner is a helper type to execute tests in different conditions. It
+// creates a go routine that is executed in a different mount namespace,
+// User ID, etc. to simulate actions hapenning inside containers.
+type Runner struct {
+	config *Config
+
+	tasks   chan func() error
+	replies chan error
+	Info    *Info
+}
+
+func NewRunner(config *Config) (*Runner, error) {
+	if config == nil {
+		config = &Config{}
+	}
+	r := &Runner{
+		config:  config,
+		tasks:   make(chan func() error),
+		replies: make(chan error),
+	}
+
+	go r.run()
+
+	if err := <-r.replies; err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
+func (r *Runner) Run(f func() error) error {
+	r.tasks <- f
+	return <-r.replies
+}
+
+func (r *Runner) Close() {
+	if r.tasks != nil {
+		close(r.tasks)
+		r.tasks = nil
+	}
+	if r.replies != nil {
+		close(r.replies)
+		r.replies = nil
+	}
+}
+
+func (r *Runner) run() {
+	// don't switch thread in the following code
+	runtime.LockOSThread()
+	// We aren't calling runtime.UnlockOSThread() to let this thread
+	// die because some changes are not recoverable, like switching
+	// to a non-root UID.
+
+	mntns, err := createMntNamespace()
+	if err != nil {
+		r.replies <- fmt.Errorf("failed to create mount namespace: %w", err)
+		return
+	}
+	defer unix.Close(mntns)
+
+	if r.config.UID != 0 {
+		// syscall.Setuid() can't be used here because it'll
+		// change the UID of all threads and we only need to
+		// change the one of this thread.
+		// https://github.com/golang/go/commit/d1b1145cace8b968307f9311ff611e4bb810710c
+		_, _, errno := syscall.Syscall(syscall.SYS_SETUID, uintptr(r.config.UID), 0, 0)
+		if errno != 0 {
+			r.replies <- fmt.Errorf("failed to set uid: %w", err)
+			return
+		}
+	}
+
+	pid := os.Getpid()
+	tid := unix.Gettid()
+
+	mountnsid, err := getMntNamespaceInode(pid, tid)
+	if err != nil {
+		r.replies <- fmt.Errorf("failed to get namespace inode: %w", err)
+		return
+	}
+
+	comm, err := os.Executable()
+	if err != nil {
+		r.replies <- fmt.Errorf("failed to get current executable: %w", err)
+	}
+
+	r.Info = &Info{
+		Pid:       pid,
+		Comm:      filepath.Base(comm),
+		UID:       r.config.UID,
+		MountNsID: mountnsid,
+	}
+
+	// Indicate it's ready to process tasks
+	r.replies <- nil
+
+	for task := range r.tasks {
+		r.replies <- task()
+	}
+}
+
+func createMntNamespace() (int, error) {
+	if err := unix.Unshare(syscall.CLONE_NEWNS); err != nil {
+		return -1, err
+	}
+	return Get()
+}
+
+func getMntNamespaceInode(pid, tid int) (uint64, error) {
+	fileinfo, err := os.Stat(fmt.Sprintf("/proc/%d/task/%d/ns/mnt", pid, tid))
+	if err != nil {
+		return 0, err
+	}
+	stat, ok := fileinfo.Sys().(*syscall.Stat_t)
+	if !ok {
+		return 0, errors.New("not a syscall.Stat_t")
+	}
+	return stat.Ino, nil
+}
+
+// Following functions have been taken from https://github.com/vishvananda/netlink
```

Comment:
It's in master -> https://github.com/vishvananda/netns/blob/50045581ed74c15eded78746bae9fb3df07d6655/netns_linux.go#L121

Reply:
ok, I used a bad terminology. Functions are not copy & paste but they have been changed to work the mount namespace instead of the network namespace.